src/main/java/org/zalando/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -317,4 +326,101 @@ public ZkSubscriptionNode getZkSubscriptionNodeLocked() {
 
         return subscriptionNode;
     }
+
+    @Override
+    public ZKSubscription subscribeForCursorsReset(final Runnable listener)
+            throws NakadiRuntimeException, UnsupportedOperationException {
+        final NodeCache cursorResetCache = new NodeCache(curatorFramework, resetCursorPath);
+        cursorResetCache.getListenable().addListener(() -> listener.run());
+
+        try {
+            cursorResetCache.start();
+        } catch (final Exception e) {
+            throw new NakadiRuntimeException(e);
+        }
+
+        return new ZKSubscription() {
+            @Override
+            public void refresh() {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public void cancel() {
+                try {
+                    cursorResetCache.getListenable().clear();
+                    cursorResetCache.close();
+                } catch (final IOException e) {
+                    throw new NakadiRuntimeException(e);
+                }
+            }
+        };
+    }
+
+    @Override
+    public boolean isCursorResetInProgress() {
+        try {
+            return curatorFramework.checkExists().forPath(resetCursorPath) != null;
+        } catch (final Exception e) {
+            // nothing in the path
+        }
+        return false;
+    }
+
+    @Override
+    public void resetCursors(final List<NakadiCursor> cursors, final Long timeout) throws OperationTimeoutException,
+            ZookeeperException, OperationInterruptedException, RequestInProgressException{
+        ZKSubscription sessionsListener = null;
+        boolean resetWasAlreadyInitiated = false;
+        try {
+            // close subscription connections
+            curatorFramework.create().withMode(CreateMode.EPHEMERAL).forPath(resetCursorPath);
+
+            final AtomicBoolean sessionsChanged = new AtomicBoolean(true);
+            sessionsListener = subscribeForSessionListChanges(() -> sessionsChanged.set(true));
+
+            final String sessionsPath = MessageFormat.format("/nakadi/subscriptions/{0}/sessions", subscriptionId);
+            final long finishAt = System.currentTimeMillis() + timeout;
+            while (finishAt > System.currentTimeMillis()) {
+                if (sessionsChanged.compareAndSet(true, false) &&
+                        curatorFramework.getChildren().forPath(sessionsPath).isEmpty()) {
+                    for (final NakadiCursor cursor : cursors) {
+                        final String path = MessageFormat
+                                .format("/nakadi/subscriptions/{0}/topics/{1}/{2}/offset",
+                                        subscriptionId, cursor.getTopic(), cursor.getPartition());
+                        curatorFramework.setData().forPath(path, cursor.getOffset().getBytes(Charsets.UTF_8));
+                    }
+                    return;
+                }
+
+                Thread.sleep(100);
```

Comment:
I know that it's boring, but maybe it's better to wait for event instead of Thread.sleep.
In this particular place one can use anything (but remember, you will live knowing that you wrote it)